### ISSUE TITLE
Calculating sensitivities in toy versus Bayesian problems

### ISSUE THREAD

**Issue by ben18785** [pints-team/pints#476]
In HMC/other-methods-needing-sensitivities, we need the derivative of the log_pdf. This can mean one of two things:

- For toy models we just need a function that calculates the sensitivities of the log_pdfs. Simple enough.
- For Bayesian problems log_pdf = log_likelihood + log_prior and, hence, when calling a function that calculates sensitivities, in this case we need log_likelihood.sensitivities + log_prior.sensitivities. So for composite log_pdfs we need a different way of calculating sensitivities.

What do people think is the best way to handle this difference? Do we need two different sub-classes of log_pdf?

Also, looking at the code, I am not sure what the value is on returning the sensitivity and additionally the log-likelihood when calling evaluateS1? In HMC, the 'ask' step will involve determining the sensitivity (and so probably we can't use ask/tell here) without the need for the log-likelihood itself, and so we are introducing additional function evaluations.

Finally, seeing as I think we will find it difficult to fit HMC into ask/tell, do we need a new class (not mcmc, so something like mcmc_sensitivity)?

**Comment by MichaelClerx** [comment]
> What do people think is the best way to handle this difference? Do we need two different sub-classes of log_pdf?

It's already implemented. LogPrior, LogLikelihood, and LogPosterior all inherit from LogPDF. In addition, several of the LogPrior implementations have an `evaluateS1` implementation, as does the LogPosterior class

**Comment by MichaelClerx** [comment]
> Also, looking at the code, I am not sure what the value is on returning the sensitivity and additionally the log-likelihood when calling evaluateS1?

I'm not sure what you mean by this!

> In HMC, the 'ask' step will involve determining the sensitivity (and so probably we can't use ask/tell here) without the need for the log-likelihood itself, and so we are introducing additional function evaluations.

Which sensitivity exactly? The sensitivity of the current position will already have been calculated in the previous tell() step

**Comment by ben18785** [comment]
> It's already implemented. LogPrior, LogLikelihood, and LogPosterior all inherit from LogPDF. In addition, several of the LogPrior implementations have an evaluateS1 implementation, as does the LogPosterior class

Great, thanks.

Re: sensitivities, the previous tell step is only at the previous position of the sampler. Not at the current intermediate position. In HMC we follow the particle along a series of small steps before we finally stop and then call tell. At each of those intermediate positions, we need a new gradient.

**Comment by ben18785** [comment]
Or are you suggesting that we call tell after each intermediate step? In other words,

- ask: if intermediate step, we take a step in parameter space according to the particle's initial momentum; if a first step we propose an initial value of the momentum and then take a step
- we calculate the sensitivity of the new position (does this go in tell?)
- tell: if an intermediate step, we always accept the position; if a terminal step we do a MH-like accept/reject

**Comment by MichaelClerx** [comment]
Would this be solved by allowing `tell` to return `None` if no new sample was accepted #320 ? In other words, is it the same issue as we have for emcee hammer #148 ? Because that's on my to-do list anyway, so could move it up a bit higher...

**Comment by ben18785** [comment]
Errm. But in the intermediate steps we ALWAYS want to accept steps, not reject them. I suspect I am being slow on the coding front here, though?

If it helps, HMC looks like this,

for 1:n_iter:
1. Sample momentum m ~ N(0, Sigma)
2. for 1:n_discrete_steps:
- Calculate sensitivity at point
- Step particle according to momentum and sensitivity to result in a new position
- Calculate momentum
3. Based on (position, momentum) after n_discrete_steps, either accept or reject new position.

**Comment by MichaelClerx** [comment]
Thanks!

But which particles are stored? Do the 1:n discrete steps end up in the chain?

**Comment by ben18785** [comment]
No, just the end positions (which are either accepted new points or the same position as at the start of each iteration).

**Comment by MichaelClerx** [comment]
So would that look like this?

```
ask() --> return Sample momentum m ~ N(0, Sigma)
tell() --> use logpdf and sens (I presume? or is this unnecessary?)
step particle
for 1:n_steps - 1:
  ask() --> ask for particle's logpdf and sensitivity
  tell() --> step particle
ask() --> ask for particle's logpdf and sensitivity
tell() --> Calculate momentum, based on (position, momentum) either accept or reject
```

so only the last tell() would return a new point for the chain, the others could return None ?

**Comment by ben18785** [comment]
Nearly... although there's no need for the logpdfs and a few of the steps.

`ask() --> return Sample momentum m ~ N(0, Sigma)
for 1:n_steps - 1:
  ask() --> ask for particle's sensitivity at new position
  tell() --> step particle according to sensitivity and momentum
  other step: calculate new momentum
ask() --> ask for particle's logpdf at new position
tell() --> Calculate momentum, based on (logpdf, momentum) either accept or reject`

**Comment by MichaelClerx** [comment]
Hmmm... In general you'd get the logpdfs for free when calculating the sensitivities, so we didn't do an `evaluate` version that provides only sensitivities (you get either only the ODE solution, or the ODE solution + derivatives). But that's ok it just means we have to ignore the logpdf when we don't need it.

For the first ask(), do we need either a sensitivity or a logpdf for this point? Otherwise it'd just become

```
Sample momentum m ~ ...
for 1:n_steps - 1:
    ask() --> ask for particle's sensitivity at new position
etc
```

**Comment by MichaelClerx** [comment]
Is this solved now @ben18785 ? Can we close this ticket?
